File: lib/js_bridge.js
// Copyright (c) 2014-2017, Jim Trainor. All rights reserved. Use of this js_bridge
// source code is governed by a BSD-style license that can be found in the LICENSE file.

/**
This is the javascript half of the Dart js_bridge package. The Dart JsBridge class
registers JS_BRIDGE Namespace instances and registers Dart function handlers in those
namespaces. Javascript application code must lookup the namespace by name and can then
use the namespace object to call Dart functions registered in that namespace.

// Lookup a namespace that is known to have been registered.
var myDartNamespace = JS_BRIDGE.lookupNamespace("myDartNamespace");

// Call a registered Dart function by name. You have to get the
// length of the argument list correct or else the Dart code will
// throw an exception.
myDartNamespace.callHandler("myDartFunction", [arg1, arg2]);

// or call the named function directly using dot notation
myDartNamespace.myDartFunction(arg1, args);
*/

/*jslint browser: true, stupid: true */

/*global JS_BRIDGE : true */

/*properties
    apply, Namespace, namespaceName, namespaces, handlers, handlerNames, isRegisteredHandler,
    getRegisteredHandlerNames, registerHandler, deregisterHandler, getHandler, callHandler,
    push, prototype, lookupNamespace, registerNamespace, readyListeners, notifyReady, addReadyListener,
    ready, length
*/

if (!JS_BRIDGE) {
    var JS_BRIDGE = {};
}

(function () {
    'use strict';

    var jsb = JS_BRIDGE;

    jsb.namespaces = {};
    jsb.readyListeners = {};

    // Notify all namespace ready listeners.
    jsb.notifyReady = function (namespaceName) {
        var i,
            listeners = jsb.readyListeners[namespaceName],
            namespace = jsb.lookupNamespace(namespaceName);
        if (listeners && namespace && namespace.ready) {
            for (i = 0; i < listeners.length; i = i + 1) {
                listeners[i](namespace);
            }
        }
    };

    // Add a callback that is called when the namespace is registered.
    // If the namespace is ready then ready listeners are called immediately.
    jsb.addReadyListener = function (namespaceName, listener) {
        if (!jsb.readyListeners[namespaceName]) {
            jsb.readyListeners[namespaceName] = [];
        }
        jsb.readyListeners[namespaceName].push(listener);
        jsb.notifyReady(namespaceName);
    };

    // Lookup and instance of jsb.Namespace by name.
    jsb.lookupNamespace = function (namespaceName) {
        return jsb.namespaces[namespaceName];
    };

    // Register an instance of jsb.Namespace.
    jsb.registerNamespace = function (namespace) {
        jsb.namespaces[namespace.namespaceName] = namespace;
    };

    // Create a js_bridge Namespace.
    jsb.Namespace = function Namespace(namespaceName) {
        this.namespaceName = namespaceName;
        this.ready = false;
        this.handlerNames = [];
    };

    jsb.Namespace.prototype.notifyReady = function () {
        this.ready = true;
        jsb.notifyReady(this.namespaceName);
    };

    // Get the list of handler names registered with this namespace.
    jsb.Namespace.prototype.getRegisteredHandlerNames = function () {
        return this.handlerNames;
    };

    // Register a function handler with this namespace. This is called by
    // the Dart JsBridge class to expose a Dart function.
    jsb.Namespace.prototype.registerHandler = function (name, handler) {
        this.handlerNames.push(name);
        this[name] = handler;
    };

    // Deregister a function handler by name.
    jsb.Namespace.prototype.deregisterHandler = function (name) {
        delete this[name];
        delete jsb.namespaces[name];
        delete jsb.readyListeners[name];
    };

    // Return true if the named function handler is registered.
    jsb.Namespace.prototype.isRegisteredHandler = function (name) {
        return this[name] !== undefined;
    };

    // Return the named function handler.
    jsb.Namespace.prototype.getHandler = function (name) {
        return this[name];
    };

    // Call the a function handler by name. This is one way for the
    // Javascript application to call handler.
    jsb.Namespace.prototype.callHandler = function (name, args) {
        return this[name].apply(this, args);
    };
}());

File: lib/js_bridge.dart
// Copyright (c) 2014, Jim Trainor. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library js_bridge;

import 'dart:js' as js;
import 'dart:async';

part 'js2dart.dart';
part 'dart2js.dart';
part 'js_callback.dart';

typedef dynamic Function0();
typedef dynamic Function1(var arg1);
typedef dynamic Function2(var arg1, var arg2);
typedef dynamic Function3(var arg1, var arg2, var arg3);
typedef dynamic Function4(var arg1, var arg2, var arg3, var arg4);
typedef dynamic Function5(var arg1, var arg2, var arg3, var arg4, var arg5);
typedef dynamic Function6(var arg1, var arg2, var arg3, var arg4, var arg5, var arg6);

_j(var arg) => js2dart(arg);
_d(var arg) => dart2js(arg);

/**
 * Expose dart functions to javascript callers.
 *
 * Registration class for dart functions that are accessible from javascript. A JsBridge
 * instance is a namespace for a collection of registered dart functions. All incoming
 * (from javascript) arguments are mapped to dart values. Primitive types (string, int,
 * double), arrays, and maps (javascript objects) map by passed as arguments. The return
 * value is mapped to a javascript primitive, array, or map. All calls registered with
 * a single [JsBridge] instance are executed in a shared error zone.
 */
class JsBridge {

  String _namespace;
  Function _onZoneError;

  js.JsObject _jsContext;

  /**
   * Create a JsBridge with the specified name name space. Optionally provide an
   * [onZoneError] function to use as the [runZoned] error handler for all registered
   * call handlers.
   */
  JsBridge(this._namespace, [Function onZoneError = print]) {
    _jsContext = new js.JsObject(js.context['JS_BRIDGE']['Namespace'], [_namespace]);
    _onZoneError = onZoneError;
    js.context['JS_BRIDGE'].callMethod('registerNamespace', [_jsContext]);
  }

  /**
   * Notify listeners that the bridge is ready. Note that listeners are registered
   * by the javascript code using the JS_BRIDGE.addRegistrationListener function.
   * Ready listeners added after this notifyReady() call are notifed at the moment
   * they are added.
   */
  void notifyReady() => _jsContext.callMethod('notifyReady');
  
  dynamic _zoned(body()) {
    return runZoned(body, onError: _onZoneError);
  }

  /**
   * Register a [handler] that takes no arguments as [name].
   */
  void registerHandler0(String name, Function0 handler) {
    makeMapper(Function0 f) => () => _zoned(()=>_d(f()));
    _register(name, makeMapper(handler));
  }

  /**
   * Register a [handler] that takes one argument as [name].
   */
  void registerHandler1(String name, Function1 handler) {
    makeMapper(Function1 f) => (a1) => _zoned(()=>_d(f(_j(a1))));
    _register(name, makeMapper(handler));
  }

  /**
   * Register a [handler] that takes two arguments as [name].
   */
  void registerHandler2(String name, Function2 handler) {
    makeMapper(Function2 f) => (a1, a2) => _zoned(()=>_d(f(_j(a1), _j(a2))));
    _register(name, makeMapper(handler));
  }

  /**
   * Register a [handler] that takes three arguments as [name].
   */
  void registerHandler3(String name, Function3 handler) {
    makeMapper(Function3 f) => (a1, a2, a3) => _zoned(()=>_d(f(_j(a1), _j(a2), _j(a3))));
    _register(name, makeMapper(handler));
  }

  /**
   * Register a [handler] that takes four arguments as [name].
   */
  void registerHandler4(String name, Function4 handler) {
    makeMapper(Function4 f) => (a1, a2, a3, a4) => _zoned(()=>_d(f(_j(a1), _j(a2), _j(a3), _j(a4))));
    _register(name, makeMapper(handler));
  }

  /**
   * Register a [handler] that takes five arguments as [name].
   */
  void registerHandler5(String name, Function5 handler) {
    makeMapper(Function5 f) => (a1, a2, a3, a4, a5) => _zoned(()=>_d(f(_j(a1), _j(a2), _j(a3), _j(a4), _j(a5))));
    _register(name, makeMapper(handler));
  }

  /**
   * Register a [handler] that takes six arguments as [name].
   */
  void registerHandler6(String name, Function6 handler) {
    makeMapper(Function6 f) => (a1, a2, a3, a4, a5, a6) => _zoned(()=>_d(f(_j(a1), _j(a2), _j(a3), _j(a4), _j(a5), _j(a6))));
    _register(name, makeMapper(handler));
  }

  void _register(name, Function handler) => _jsContext.callMethod('registerHandler', [name, handler]);

  /**
   * Deregister a handler by [name].
   */
  void deregisterHandler(String name) {
    _jsContext.callMethod('deregisterHandler', [name]);
  }

  /**
   * Return true if a handler with [name] is registered.
   */
  bool isRegisteredHandler(String name) {
    return _jsContext.callMethod('isRegisteredHandler', [name]);
  }
}
